enterprise/server/util/chunker/chunker.go

```diff
@@ -0,0 +1,122 @@
+package chunker
+
+import (
+	"context"
+	"io"
+
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/jotfs/fastcdc-go"
+)
+
+type WriteFunc func([]byte) error
+
+type Chunker struct {
+	ctx context.Context
+	pw  *io.PipeWriter
+	pr  *io.PipeReader
+
+	done      chan struct{}
+	closed    bool
+	numChunks int
+
+	singleChunkWriteFn   WriteFunc
+	multipleChunkWriteFn WriteFunc
+}
+
+func (c *Chunker) Write(buf []byte) (int, error) {
+	return c.pw.Write(buf)
+}
+
+func (c *Chunker) Close() error {
+	if err := c.pw.Close(); err != nil {
+		return err
+	}
+
+	select {
+	case <-c.ctx.Done():
+		return c.ctx.Err()
+	case <-c.done:
+		break
+	}
+	c.closed = true
+	return nil
+}
+
+func (c *Chunker) NumChunks() (int, error) {
+	if !c.closed {
+		return 0, status.InternalErrorf("cannot call NumChunks() before chunker is closed")
+	}
+	return c.numChunks, nil
+}
+
+// New returns an io.WriteCloser that split file into chunks of average size.
+// averageSize is typically a power of 2. It must be in the range 64B to 1GiB.
+// The minimum allowed chunk size is averageSize / 4, and the maximum allowed
+// chunk size is averageSize * 4.
+//
+// singleChunkWriteFn and multipleChunkWriteFn specify how to write raw chunk
+// when there is only one chunk and more than one chunks respectively
+func New(ctx context.Context, averageSize int, singleChunkWriteFn WriteFunc, multipleChunkWriteFn WriteFunc) (*Chunker, error) {
+	pr, pw := io.Pipe()
+	c := &Chunker{
+		ctx:                  ctx,
+		pr:                   pr,
+		pw:                   pw,
+		done:                 make(chan struct{}, 0),
+		singleChunkWriteFn:   singleChunkWriteFn,
+		multipleChunkWriteFn: multipleChunkWriteFn,
+	}
+	cdcOpts := fastcdc.Options{
```

Comment:
nit: could you write a comment here to explain why these are the options you picked and if they would ever change?

Reply:
I was using the default min and max; i don't think i need to specify them.

Reply:
might be worth specifying these still just in case the underlying library decides to change them?

Reply:
Done.